core/src/autogluon/core/trainer/abstract_trainer.py

```diff
@@ -591,13 +605,17 @@ def get_model_pred_proba_dict(self, X, models, model_pred_proba_dict=None, model
         if fit:
             model_pred_order = [model for model in models if model not in model_pred_proba_dict.keys()]
         else:
+            if use_val_cache:
+                _, model_pred_proba_dict = self._update_pred_proba_dict_with_val_cache(model_set=set(models), model_pred_proba_dict=model_pred_proba_dict)
             model_set = set()
             for model in models:
-                if model in model_set:
+                if model in model_set or model in model_pred_proba_dict:
                     continue
                 min_model_set = set(self.get_minimum_model_set(model))
                 model_set = model_set.union(min_model_set)
             model_set = model_set.difference(set(model_pred_proba_dict.keys()))
+            if use_val_cache:
+                model_set, model_pred_proba_dict = self._update_pred_proba_dict_with_val_cache(model_set=model_set, model_pred_proba_dict=model_pred_proba_dict)
```

Comment:
Can you explain why we are calling `_update_pred_proba_dict_with_val_cache` both here and line 609?

Reply:
We don't know the full DAG of models we need to predict with yet at 609. 609 is used to filter the head of the DAG so we don't need to consider models that are redundant because the head has been pruned. In the case of a many-level stack, this reduces complexity from O(n) to O(1).

In case the heads aren't cached but some DAG members are, this is where 618 fetches these cases.

Essentially we call twice for the purposes of compute optimization.

Reply:
makes sense